drivers/r2dbc-driver/src/main/kotlin/app/cash/sqldelight/driver/r2dbc/R2dbcDriver.kt

```diff
@@ -0,0 +1,195 @@
+package app.cash.sqldelight.driver.r2dbc
+
+import app.cash.sqldelight.async.AsyncQuery
+import app.cash.sqldelight.async.AsyncTransacter
+import app.cash.sqldelight.async.db.AsyncSqlCursor
+import app.cash.sqldelight.async.db.AsyncSqlDriver
+import app.cash.sqldelight.async.db.AsyncSqlPreparedStatement
+import io.r2dbc.spi.Connection
+import io.r2dbc.spi.Statement
+import kotlinx.coroutines.reactive.awaitLast
+import kotlinx.coroutines.reactive.awaitSingle
+import org.reactivestreams.Publisher
+import org.reactivestreams.Subscriber
+import org.reactivestreams.Subscription
+
+class R2dbcDriver(val connection: Connection) : AsyncSqlDriver {
+  override suspend fun <R> executeQuery(
+    identifier: Int?,
+    sql: String,
+    mapper: (AsyncSqlCursor) -> R,
+    parameters: Int,
+    binders: (AsyncSqlPreparedStatement.() -> Unit)?
+  ): R {
+    val prepared = connection.createStatement(sql).also { statement ->
+      R2dbcPreparedStatement(statement).apply { if (binders != null) this.binders() }
+    }
+    val result = prepared.execute().awaitSingle()
+
+    val rowSet = mutableListOf<Map<Int, Any?>>()
+    result.map { row, rowMetadata ->
+      rowSet.add(rowMetadata.columnMetadatas.mapIndexed { index, _ -> index to row.get(index) }.toMap())
+    }.awaitLast()
+
+    return mapper(R2dbcCursor(rowSet))
+  }
+
+  override suspend fun execute(
+    identifier: Int?,
+    sql: String,
+    parameters: Int,
+    binders: (AsyncSqlPreparedStatement.() -> Unit)?
+  ): Long {
+    val prepared = connection.createStatement(sql).also { statement ->
+      R2dbcPreparedStatement(statement).apply { if (binders != null) this.binders() }
+    }
+
+    val result = prepared.execute().awaitSingle()
+    // TODO: r2dbc-mysql emits a java.lang.Integer instead of a java.lang.Long, mysql driver needs to support latest r2dbc-spi
+    // return result.rowsUpdated.awaitSingle()
+    return 0L
```

Comment:
An older version of the R2DBC SPI has the `execute()` method returning an `Int`, but in a newer version it was changed to a `Long`. The latest release of the MySQL driver implements the older version of the SPI so there's an ABI incompatibility that we can't do anything about.

I think I'm going to get rid of that comment and return the actual `rowsUpdated` result since this is a weird case to be handling, but we'll be left in a funny situation where there are zero R2DBC drivers that this SQLDelight driver can actually work with (as-is anyway). lol...

Reply:
oh dear okay. Do you mean that `rowsUpdated` changed from an Int to a Long? Or do both old and new versions correctly support rowsUpdated so that would be fine

Reply:
It changed from an Int to a Long, and the mysql driver just hasn't seen a new release in a long time (no pun intended)